golang/internal/runtime/container/container.go

```diff
@@ -0,0 +1,120 @@
+package containerruntime
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/docker/docker/client"
+	"github.com/hashicorp/go-version"
+	"github.com/rs/zerolog/log"
+)
+
+// If a runtime version doesn't satisfy recommendation: show warning
+// If a runtime version doesn't satisfy minimum: failure\
+
+const (
+	Podman                         = "podman"
+	Docker                         = "docker"
+	MinimumDockerServerVersion     = "20.10.0"
+	RecommendedDockerServerVersion = "23.0.0"
+	MinimumPodmanServerVersion     = "4.0.0"
+	RecommendedPodmanServerVersion = "4.4.0"
+	PodmanHost                     = "host.containers.internal"
+	DockerHost                     = "host.docker.internal"
+)
+
+func VersionCheck() {
+	ctx := context.Background()
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		log.Fatal().Stack().Err(err).Msg("Server connection error")
+	}
+
+	serverVersion, err := cli.ServerVersion(ctx)
+	if err != nil {
+		log.Fatal().Stack().Err(err).Msg("Could not get server version")
+	}
+
+	switch getContainerRuntime() {
+	case Podman:
+		log.Info().Str("version", serverVersion.Version).Msg("podman")
+
+		if !satisfyVersion(MinimumPodmanServerVersion, serverVersion.Version) {
+			log.Fatal().
+				Str("serverVersion", serverVersion.Version).
+				Str("minVersion", MinimumPodmanServerVersion).
+				Msg("Server is not a supported version")
+		}
+		if !satisfyVersion(RecommendedPodmanServerVersion, serverVersion.Version) {
+			log.Warn().
+				Str("serverVersion", serverVersion.Version).
+				Str("reqVersion", RecommendedPodmanServerVersion).
+				Msg("Server is behind the supported version")
+		}
+	case Docker:
+		log.Info().Str("version", serverVersion.Version).Msg("Docker")
+		if !satisfyVersion(MinimumDockerServerVersion, serverVersion.Version) {
+			log.Fatal().
+				Str("serverVersion", serverVersion.Version).
+				Str("minVersion", MinimumDockerServerVersion).
+				Msg("Server is not a supported version")
+		}
+		if !satisfyVersion(RecommendedPodmanServerVersion, serverVersion.Version) {
+			log.Warn().
+				Str("serverVersion", serverVersion.Version).
+				Str("reqVersion", RecommendedDockerServerVersion).
+				Msg("Server is behind the supported version")
+		}
+
+	default:
+		log.Fatal().Msg("Unknown container runtime")
+	}
+}
+
+func satisfyVersion(preferredVer, actualVer string) bool {
+	serVer, err := version.NewVersion(actualVer)
+	if err != nil {
+		log.Fatal().Stack().Err(err).Msg("Invalid version string from server")
+	}
+
+	constraints, err := version.NewConstraint(fmt.Sprintf(">=%s", preferredVer))
+	if err != nil {
+		log.Error().Stack().Err(err).Msg("Error with version constraint")
+	}
+
+	if constraints.Check(serVer) {
+		return true
+	}
+	return false
+}
+
+func GetInternalHostDomain() string {
+	containerRuntime := getContainerRuntime()
+
+	if containerRuntime == Podman {
+		return PodmanHost
+	}
+	return DockerHost
+}
+
+func getContainerRuntime() string {
```

Comment:
spawning a client/ctx in function makes testing a lot complicated, use parameters.